False positive `used-before-assignment` for homonym between subscript in comprehension and assignment in except block

False positive for `used-before-assignment` from subscript notation in comprehension followed by definition of that same variable later in an except block.

```python
# test.py
def select_dict(data, keys):
    try:
        d = {key: data[key] for key in keys}
    except KeyError as e:
        key, *_ = e.args
        raise Exception(f"{key} not found")

    return d
```

```console
$ pip install -U pylint==2.13.4 -q && pylint test.py
************* Module test
test.py:3:23: E0601: Using variable 'key' before assignment (used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 2.86/10 (previous run: 2.86/10, +0.00)
```

_Originally posted by @skshetry in https://github.com/PyCQA/pylint/issues/6035#issuecomment-1084339275_

--- a/ChangeLog
+++ b/ChangeLog
@@ -41,6 +41,11 @@ What's New in Pylint 2.13.5?
 ============================
 Release date: TBA
 
+* Fix false positive regression in 2.13.0 for ``used-before-assignment`` for
+  homonyms between variable assignments in try/except blocks and variables in
+  subscripts in comprehensions.
+
+  Closes #6069
 
 
 What's New in Pylint 2.13.4?

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1520,8 +1520,8 @@ class VariablesChecker(BaseChecker):
         if node.name in current_consumer.consumed:
             # Avoid the case where there are homonyms inside function scope and
             # comprehension current scope (avoid bug #1731)
-            if utils.is_func_decorator(current_consumer.node) or not (
-                isinstance(node, nodes.ComprehensionScope)
+            if utils.is_func_decorator(current_consumer.node) or not isinstance(
+                node, nodes.ComprehensionScope
             ):
                 self._check_late_binding_closure(node)
                 self._loopvar_name(node)

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1513,7 +1513,13 @@ class VariablesChecker(BaseChecker):
                         isinstance(node.scope(), nodes.ComprehensionScope)
                         and isinstance(
                             node.parent,
-                            (nodes.Call, nodes.Compare, nodes.Keyword, nodes.Subscript),
+                            (
+                                nodes.Call,
+                                nodes.BaseContainer,
+                                nodes.Compare,
+                                nodes.Keyword,
+                                nodes.Subscript,
+                            ),
                         )
                     )
                 )

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1281,21 +1281,13 @@ class VariablesChecker(BaseChecker):
 
         global_names = _flattened_scope_names(node.nodes_of_class(nodes.Global))
         nonlocal_names = _flattened_scope_names(node.nodes_of_class(nodes.Nonlocal))
-        comprehension_target_names = []
-
-        def find_assigned_names_recursive(
-            target: Union[nodes.AssignName, nodes.BaseContainer]
-        ):
-            nonlocal comprehension_target_names
-            if isinstance(target, nodes.AssignName):
-                comprehension_target_names.append(target.name)
-            elif isinstance(target, nodes.BaseContainer):
-                for elt in target.elts:
-                    find_assigned_names_recursive(elt)
+        comprehension_target_names: List[str] = []
 
         for comprehension_scope in node.nodes_of_class(nodes.ComprehensionScope):
             for generator in comprehension_scope.generators:
-                find_assigned_names_recursive(generator.target)
+                self._find_assigned_names_recursive(
+                    generator.target, comprehension_target_names
+                )
 
         for name, stmts in not_consumed.items():
             self._check_is_unused(
@@ -1503,6 +1495,20 @@ class VariablesChecker(BaseChecker):
             return True
 
         return False
+
+    def _find_assigned_names_recursive(
+        self,
+        target: Union[nodes.AssignName, nodes.BaseContainer],
+        target_names: List[str],
+    ) -> None:
+        """Update `target_names` in place with the names of assignment
+        targets, recursively (to account for nested assignments).
+        """
+        if isinstance(target, nodes.AssignName):
+            target_names.append(target.name)
+        elif isinstance(target, nodes.BaseContainer):
+            for elt in target.elts:
+                self._find_assigned_names_recursive(elt, target_names)
 
     # pylint: disable=too-many-return-statements
     def _check_consumer(
@@ -2279,7 +2285,7 @@ class VariablesChecker(BaseChecker):
         stmt,
         global_names,
         nonlocal_names: Iterable[str],
-        comprehension_target_names: Iterable[str],
+        comprehension_target_names: List[str],
     ):
         # Ignore some special names specified by user configuration.
         if self._is_name_ignored(stmt, name):

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1512,7 +1512,8 @@ class VariablesChecker(BaseChecker):
                     or (
                         isinstance(node.scope(), nodes.ComprehensionScope)
                         and isinstance(
-                            node.parent, (nodes.Call, nodes.Keyword, nodes.Subscript)
+                            node.parent,
+                            (nodes.Call, nodes.Compare, nodes.Keyword, nodes.Subscript),
                         )
                     )
                 )

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1508,9 +1508,12 @@ class VariablesChecker(BaseChecker):
                     self._has_homonym_in_comprehension_test(node)
                     # Or homonyms against values to keyword arguments
                     # (like "var" in "[func(arg=var) for var in expr()]")
+                    # (or "var" in "{var: data[var] for var in expr()}")
                     or (
                         isinstance(node.scope(), nodes.ComprehensionScope)
-                        and isinstance(node.parent, (nodes.Call, nodes.Keyword))
+                        and isinstance(
+                            node.parent, (nodes.Call, nodes.Keyword, nodes.Subscript)
+                        )
                     )
                 )
             ):

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -629,8 +629,8 @@ scope_type : {self._atomic.scope_type}
         ):
             return found_nodes
 
-        # And is not part of a test in a filtered comprehension
-        if VariablesChecker._has_homonym_in_comprehension_test(node):
+        # And no comprehension is under the node's frame
+        if VariablesChecker._comprehension_under_frame(node):
             return found_nodes
 
         # Filter out assignments in ExceptHandlers that node is not contained in
@@ -1525,7 +1525,7 @@ class VariablesChecker(BaseChecker):
                 current_consumer.scope_type == "comprehension"
                 and self._has_homonym_in_upper_function_scope(node, consumer_level)
                 and not any(
-                    self._has_homonym_in_comprehension_test(comprehension)
+                    self._comprehension_under_frame(comprehension)
                     for comprehension in (node, *node.scope().generators)
                 )
             ):
@@ -2501,23 +2501,8 @@ class VariablesChecker(BaseChecker):
         )
 
     @staticmethod
-    def _has_homonym_in_comprehension_test(node: nodes.Name) -> bool:
-        """Return True if `node`'s frame contains a comprehension employing an
-        identical name in a test.
-
-        The name in the test could appear at varying depths, and in the `if` test or not:
-
-        Examples:
-            [x for x in range(3) if x]
-            [x for x in range(3) if x.num == 1]
-            [x for x in range(3)] if call(x.num)]
-            [0 | x for x in range(3)]
-
-        See:
-        https://github.com/PyCQA/pylint/issues/5586
-        https://github.com/PyCQA/pylint/issues/6035
-        https://github.com/PyCQA/pylint/issues/6069
-        """
+    def _comprehension_under_frame(node: nodes.Name) -> bool:
+        """Return True if `node`'s frame contains a ComprehensionScope."""
         closest_comprehension_scope = utils.get_node_first_ancestor_of_type(
             node, nodes.ComprehensionScope
         )

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2286,7 +2286,7 @@ class VariablesChecker(BaseChecker):
         global_names,
         nonlocal_names: Iterable[str],
         comprehension_target_names: List[str],
-    ):
+    ) -> None:
         # Ignore some special names specified by user configuration.
         if self._is_name_ignored(stmt, name):
             return

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1434,7 +1434,7 @@ class VariablesChecker(BaseChecker):
                 continue
 
             action, nodes_to_consume = self._check_consumer(
-                node, stmt, frame, current_consumer, i, base_scope_type
+                node, stmt, frame, current_consumer, base_scope_type
             )
             if nodes_to_consume:
                 # Any nodes added to consumed_uncertain by get_next_to_consume()
@@ -1512,22 +1512,16 @@ class VariablesChecker(BaseChecker):
         stmt: nodes.NodeNG,
         frame: nodes.LocalsDictNodeNG,
         current_consumer: NamesConsumer,
-        consumer_level: int,
         base_scope_type: Any,
     ) -> Tuple[VariableVisitConsumerAction, Optional[List[nodes.NodeNG]]]:
         """Checks a consumer for conditions that should trigger messages."""
         # If the name has already been consumed, only check it's not a loop
         # variable used outside the loop.
-        # Avoid the case where there are homonyms inside function scope and
-        # comprehension current scope (avoid bug #1731)
         if node.name in current_consumer.consumed:
+            # Avoid the case where there are homonyms inside function scope and
+            # comprehension current scope (avoid bug #1731)
             if utils.is_func_decorator(current_consumer.node) or not (
-                current_consumer.scope_type == "comprehension"
-                and self._has_homonym_in_upper_function_scope(node, consumer_level)
-                and not any(
-                    self._comprehension_under_frame(comprehension)
-                    for comprehension in (node, *node.scope().generators)
-                )
+                isinstance(node, nodes.ComprehensionScope)
             ):
                 self._check_late_binding_closure(node)
                 self._loopvar_name(node)
@@ -2482,23 +2476,6 @@ class VariablesChecker(BaseChecker):
 
     def _allowed_redefined_builtin(self, name):
         return name in self.config.allowed_redefined_builtins
-
-    def _has_homonym_in_upper_function_scope(
-        self, node: nodes.Name, index: int
-    ) -> bool:
-        """Return whether there is a node with the same name in the
-        to_consume dict of an upper scope and if that scope is a function
-
-        :param node: node to check for
-        :param index: index of the current consumer inside self._to_consume
-        :return: True if there is a node with the same name in the
-                 to_consume dict of an upper scope and if that scope
-                 is a function, False otherwise
-        """
-        return any(
-            _consumer.scope_type == "function" and node.name in _consumer.to_consume
-            for _consumer in self._to_consume[index - 1 :: -1]
-        )
 
     @staticmethod
     def _comprehension_under_frame(node: nodes.Name) -> bool:

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -630,7 +630,7 @@ scope_type : {self._atomic.scope_type}
             return found_nodes
 
         # And no comprehension is under the node's frame
-        if VariablesChecker._comprehension_under_frame(node):
+        if VariablesChecker._comprehension_between_frame_and_node(node):
             return found_nodes
 
         # Filter out assignments in ExceptHandlers that node is not contained in
@@ -2478,8 +2478,8 @@ class VariablesChecker(BaseChecker):
         return name in self.config.allowed_redefined_builtins
 
     @staticmethod
-    def _comprehension_under_frame(node: nodes.Name) -> bool:
-        """Return True if `node`'s frame contains a ComprehensionScope."""
+    def _comprehension_between_frame_and_node(node: nodes.Name) -> bool:
+        """Return True if a ComprehensionScope intervenes between `node` and its frame."""
         closest_comprehension_scope = utils.get_node_first_ancestor_of_type(
             node, nodes.ComprehensionScope
         )

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1282,8 +1282,31 @@ class VariablesChecker(BaseChecker):
 
         global_names = _flattened_scope_names(node.nodes_of_class(nodes.Global))
         nonlocal_names = _flattened_scope_names(node.nodes_of_class(nodes.Nonlocal))
+        comprehension_target_names = []
+
+        def find_assigned_names_recursive(
+            target: Union[nodes.AssignName, nodes.BaseContainer]
+        ):
+            nonlocal comprehension_target_names
+            if isinstance(target, nodes.AssignName):
+                comprehension_target_names.append(target.name)
+            elif isinstance(target, nodes.BaseContainer):
+                for elt in target.elts:
+                    find_assigned_names_recursive(elt)
+
+        for comprehension_scope in node.nodes_of_class(nodes.ComprehensionScope):
+            for generator in comprehension_scope.generators:
+                find_assigned_names_recursive(generator.target)
+
         for name, stmts in not_consumed.items():
-            self._check_is_unused(name, node, stmts[0], global_names, nonlocal_names)
+            self._check_is_unused(
+                name,
+                node,
+                stmts[0],
+                global_names,
+                nonlocal_names,
+                comprehension_target_names,
+            )
 
     visit_asyncfunctiondef = visit_functiondef
     leave_asyncfunctiondef = leave_functiondef
@@ -1501,28 +1524,9 @@ class VariablesChecker(BaseChecker):
             if utils.is_func_decorator(current_consumer.node) or not (
                 current_consumer.scope_type == "comprehension"
                 and self._has_homonym_in_upper_function_scope(node, consumer_level)
-                # But don't catch homonyms against the filter of a comprehension,
-                # (like "if x" in "[x for x in expr() if x]")
-                # https://github.com/PyCQA/pylint/issues/5586
-                and not (
-                    self._has_homonym_in_comprehension_test(node)
-                    # Or homonyms against values to keyword arguments
-                    # (like "var" in "[func(arg=var) for var in expr()]")
-                    # (or "var" in "{var: data[var] for var in expr()}")
-                    or (
-                        isinstance(node.scope(), nodes.ComprehensionScope)
-                        and isinstance(
-                            node.parent,
-                            (
-                                nodes.BaseContainer,
-                                nodes.BinOp,
-                                nodes.Call,
-                                nodes.Compare,
-                                nodes.Keyword,
-                                nodes.Subscript,
-                            ),
-                        )
-                    )
+                and not any(
+                    self._has_homonym_in_comprehension_test(comprehension)
+                    for comprehension in (node, *node.scope().generators)
                 )
             ):
                 self._check_late_binding_closure(node)
@@ -2276,7 +2280,13 @@ class VariablesChecker(BaseChecker):
                 self.add_message("undefined-loop-variable", args=node.name, node=node)
 
     def _check_is_unused(
-        self, name, node, stmt, global_names, nonlocal_names: Iterable[str]
+        self,
+        name,
+        node,
+        stmt,
+        global_names,
+        nonlocal_names: Iterable[str],
+        comprehension_target_names: Iterable[str],
     ):
         # Ignore some special names specified by user configuration.
         if self._is_name_ignored(stmt, name):
@@ -2299,7 +2309,8 @@ class VariablesChecker(BaseChecker):
         argnames = node.argnames()
         # Care about functions with unknown argument (builtins)
         if name in argnames:
-            self._check_unused_arguments(name, node, stmt, argnames, nonlocal_names)
+            if name not in comprehension_target_names:
+                self._check_unused_arguments(name, node, stmt, argnames, nonlocal_names)
         else:
             if stmt.parent and isinstance(
                 stmt.parent, (nodes.Assign, nodes.AnnAssign, nodes.Tuple)
@@ -2494,24 +2505,25 @@ class VariablesChecker(BaseChecker):
         """Return True if `node`'s frame contains a comprehension employing an
         identical name in a test.
 
-        The name in the test could appear at varying depths:
+        The name in the test could appear at varying depths, and in the `if` test or not:
 
         Examples:
-            [x for x in range(3) if name]
-            [x for x in range(3) if name.num == 1]
-            [x for x in range(3)] if call(name.num)]
+            [x for x in range(3) if x]
+            [x for x in range(3) if x.num == 1]
+            [x for x in range(3)] if call(x.num)]
+            [0 | x for x in range(3)]
+
+        See:
+        https://github.com/PyCQA/pylint/issues/5586
+        https://github.com/PyCQA/pylint/issues/6035
+        https://github.com/PyCQA/pylint/issues/6069
         """
-        closest_comprehension = utils.get_node_first_ancestor_of_type(
-            node, nodes.Comprehension
+        closest_comprehension_scope = utils.get_node_first_ancestor_of_type(
+            node, nodes.ComprehensionScope
         )
-        return (
-            closest_comprehension is not None
-            and node.frame(future=True).parent_of(closest_comprehension)
-            and any(
-                test is node or test.parent_of(node)
-                for test in closest_comprehension.ifs
-            )
-        )
+        return closest_comprehension_scope is not None and node.frame(
+            future=True
+        ).parent_of(closest_comprehension_scope)
 
     def _store_type_annotation_node(self, type_annotation):
         """Given a type annotation, store all the name nodes it refers to."""

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1514,8 +1514,9 @@ class VariablesChecker(BaseChecker):
                         and isinstance(
                             node.parent,
                             (
-                                nodes.Call,
                                 nodes.BaseContainer,
+                                nodes.BinOp,
+                                nodes.Call,
                                 nodes.Compare,
                                 nodes.Keyword,
                                 nodes.Subscript,

--- a/tests/functional/u/unused/unused_argument.py
+++ b/tests/functional/u/unused/unused_argument.py
@@ -47,6 +47,12 @@ def metadata_from_dict(key):
     """
     return {key: str(value) for key, value in key.items()}
 
+
+def metadata_from_dict_2(key):
+    """Similar, but with more nesting"""
+    return {key: (a, b) for key, (a, b) in key.items()}
+
+
 # pylint: disable=too-few-public-methods,  misplaced-future,wrong-import-position
 from __future__ import print_function
 

--- a/tests/functional/u/unused/unused_argument.txt
+++ b/tests/functional/u/unused/unused_argument.txt
@@ -1,9 +1,9 @@
 unused-argument:3:16:3:21:test_unused:Unused argument 'first':HIGH
 unused-argument:3:23:3:29:test_unused:Unused argument 'second':HIGH
 unused-argument:32:29:32:32:Sub.newmethod:Unused argument 'aay':INFERENCE
-unused-argument:54:13:54:16:function:Unused argument 'arg':HIGH
-unused-argument:61:21:61:24:AAAA.method:Unused argument 'arg':INFERENCE
-unused-argument:68:0:None:None:AAAA.selected:Unused argument 'args':INFERENCE
-unused-argument:68:0:None:None:AAAA.selected:Unused argument 'kwargs':INFERENCE
-unused-argument:87:23:87:26:BBBB.__init__:Unused argument 'arg':INFERENCE
-unused-argument:98:34:98:39:Ancestor.set_thing:Unused argument 'other':INFERENCE
+unused-argument:60:13:60:16:function:Unused argument 'arg':HIGH
+unused-argument:67:21:67:24:AAAA.method:Unused argument 'arg':INFERENCE
+unused-argument:74:0:None:None:AAAA.selected:Unused argument 'args':INFERENCE
+unused-argument:74:0:None:None:AAAA.selected:Unused argument 'kwargs':INFERENCE
+unused-argument:93:23:93:26:BBBB.__init__:Unused argument 'arg':INFERENCE
+unused-argument:104:34:104:39:Ancestor.set_thing:Unused argument 'other':INFERENCE

--- a/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
+++ b/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
@@ -72,3 +72,14 @@ def func7():
     except ZeroDivisionError:
         i = None
         print(i, bools)
+
+
+def func8():
+    """Similar, but with a container"""
+    pairs = [(i, i) for i in range(3)]
+
+    try:
+        1 / 0
+    except ZeroDivisionError:
+        i = None
+        print(i, pairs)

--- a/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
+++ b/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
@@ -83,3 +83,14 @@ def func8():
     except ZeroDivisionError:
         i = None
         print(i, pairs)
+
+
+def func9():
+    """Similar, but with a binary op"""
+    bools = [str(i) | i for i in range(3)]
+
+    try:
+        1 / 0
+    except ZeroDivisionError:
+        i = None
+        print(i, bools)

--- a/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
+++ b/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
@@ -49,3 +49,15 @@ def func5():
     except ZeroDivisionError:
         k = None
         print(k, filtered)
+
+
+def func6(data, keys):
+    """Similar, but with a subscript in a key-value pair rather than the test
+    See https://github.com/PyCQA/pylint/issues/6069"""
+    try:
+        results = {key: data[key] for key in keys}
+    except KeyError as exc:
+        key, *_ = exc.args
+        raise Exception(f"{key} not found") from exc
+
+    return results

--- a/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
+++ b/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
@@ -85,12 +85,11 @@ def func8():
         print(i, pairs)
 
 
-def func9():
-    """Similar, but with a binary op"""
-    bools = [str(i) | i for i in range(3)]
+# Module level cases
 
-    try:
-        1 / 0
-    except ZeroDivisionError:
-        i = None
-        print(i, bools)
+module_ints = [j | j for j in range(3)]
+try:
+    1 / 0
+except ZeroDivisionError:
+    j = None
+    print(j, module_ints)

--- a/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
+++ b/tests/functional/u/used/used_before_assignment_comprehension_homonyms.py
@@ -61,3 +61,14 @@ def func6(data, keys):
         raise Exception(f"{key} not found") from exc
 
     return results
+
+
+def func7():
+    """Similar, but with a comparison"""
+    bools = [str(i) == i for i in range(3)]
+
+    try:
+        1 / 0
+    except ZeroDivisionError:
+        i = None
+        print(i, bools)

